### PR TITLE
feat: store receipts for local pending block

### LINKED ISSUE #9948: store receipts for local pending block
### Describe the feature

currently we don't store the receipts for the pending block we just built

https://github.com/paradigmxyz/reth/blob/750eb26256b6fcea0ad378a8cf08fe28f2dc0a7b/crates/rpc/rpc-eth-types/src/pending_block.rs#L118-L118

and for eth_getBlockreceipts we don't check the local block 

https://github.com/paradigmxyz/reth/blob/750eb26256b6fcea0ad378a8cf08fe28f2dc0a7b/crates/rpc/rpc-eth-api/src/helpers/block.rs#L154-L154

## TODO
* store receipts for locally built block and integrate in `LoadPendingBlock`
 

cc @loocapro 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by malik672**
this pr closes #9948

### CODE REVIEW

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:76]: please revert unrelated changes. we use the nightly formatter in reth `cargo +nightly fmt`. try command `make lint`.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:352]:
```suggestion
                            continue
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:327]:
```suggestion
                    continue
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:311]:
```suggestion
                continue
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:303]:
```suggestion
                continue
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:175]:
```suggestion
                    return Ok(None)
```

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:174]:
```suggestion
                    .map_err(Self::Error::from_eth_err)
```

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:136]:
```suggestion
                return receipts.map(Some)
```

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:356]:
```suggestion
                            return Err(Self::Error::from_evm_err(err))
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:139]: another error would be more suitable here, since this only gets called in case `pending.origin.into_actual_pending()` actually returns `Some(block)`

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:126]: let's not change the signature here. the pending block receipts field can be read inside scope of `load_block_and_receipts`. that way you also save the clone on the receipts which you are doing now to write it to the pending block.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (malik672) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:157]: pending_block does not have the method

File: crates/rpc/rpc-eth-types/src/pending_block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-types/src/pending_block.rs:126]: hmm, do we need to store Option here?
ideally this is just a Vec

perhaps this is due to some executor API? if so we should do the flatten + collect when we set this

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by fgimenez — APPROVED** [review]
lgtm, pending @emhane and @mattsse

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:178]: this doesn't really work because we're trying to fetch a block by number that doesn't exist.

what we need to do is change how this behaves:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-api/src/helpers/pending_block.rs#L219-L228

this should also return the receipts

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:174]: here we should also change what local_pending_block returns and also make it return the receipts,

perhaps even the entire

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-api/src/helpers/pending_block.rs#L180-L180

type

**Review by mattsse — CHANGES_REQUESTED** [review]
last suggestions

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (malik672) [reply crates/rpc/rpc-eth-api/src/helpers/block.rs:174]: I did that before .....

Comment (mattsse) [reply crates/rpc/rpc-eth-api/src/helpers/block.rs:174]: I see, sorry about this...

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:76]: please undo
```suggestion
                    .map(|block| block.body.len()))
```

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:136]:
```suggestion
                return receipts.map(Some)
```

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:174]:
```suggestion
                    .map_err(Self::Error::from_eth_err)
```

File: crates/rpc/rpc-eth-api/src/helpers/block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/block.rs:263]:
```suggestion
                        Some((block, _)) => Ok(Some(block)),
```

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:140]: this is also tricky, because this can end up not returning the receipts

ideally we also integrate the receipts into

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-types/src/pending_block.rs#L61-L62

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-api/src/helpers/pending_block.rs#L73-L73

**Review by mattsse — CHANGES_REQUESTED** [review]
last suggestions,

please avoid opening prs from your main branch because maintainers can't make edits before merging.

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (malik672) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:140]: I see it tbh

Comment (malik672) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:140]: can you explain further

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:189]: If the pending block is built with receipts now, it shouldn't be necessary to return the receipts separately, so this clone should be avoidable, no? Ref https://github.com/paradigmxyz/reth/pull/10070#discussion_r1703342665

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (malik672) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:189]: I do not see a way to avoid the clones tbh, unless I change the signature of the func

**Comment by malik672** [comment]
@mattsse

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (emhane) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:189]: ah yeah see what you mean. return `&Mutex<Option<PendingBlock>>` instead of `SealedBlockWithSenders`.

Comment (emhane) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:189]: or don't return any data at all rather, so `impl Future<Output = Result<(), Self::Error>> + Send`. then call `self.pending_block()` again to read the updated pending block.

**Comment by mattsse** [comment]
could you please resolve the smol conflict, I believe this is complete now

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (mattsse) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:189]: no mutex in function signatures, this clone is appropriate

**Comment by malik672** [comment]
> could you please resolve the smol conflict, I believe this is complete now

@mattsse done

**Comment by mattsse** [comment]
@malik672 could you please check if you see an option to "allow edits by maintainers"?

**Comment by malik672** [comment]
> @malik672 could you please check if you see an option to "allow edits by maintainers"?

@mattsse done

**Review by emhane — APPROVED** [review]
can optimise this in follow up pr https://github.com/paradigmxyz/reth/pull/10070#discussion_r1709232254

**Comment by mattsse** [comment]
please merge branch #10589 into this branch so I don't have to review your changes again

so please undo your last commits and use the commits from #10589

**Comment by mattsse** [comment]
please undo the last 3 commits,

if it's easier consider reopening this pr from a different branch so I can make modifications

**Pull request closed without merging**
